### PR TITLE
Outbound split and move into folder

### LINKED ISSUE #1831: Split up outbound.js
We need to modularize outbound.js a bit more. There's several components it could be split into:

* a queue loader/processor
* a "main" library that does the main outbound bits (send_email etc)
* The outbound connection pool code
* the ToDo library
* the HMailItem library

### PR DESCRIPTION
**Pull request by baudehlo**
Fixes #1831 

Changes proposed in this pull request:
- Split outbound.js into a sub-directory and multiple files
- Fix tests to cope with this

Checklist:
- [ ] docs updated
- [X] tests updated

### CODE REVIEW

**Comment by baudehlo** [comment]
There's still more to do on this branch. Just wanted the PR in place first.

**Comment by baudehlo** [comment]
@msimerson @gwieshammer outbound with the splitting now works but the tests fail. The outbound tests are a bit of a twisty maze of passages that I'm really unfamiliar with - you two have done all the previous work on them - any chance one of you could take a peek?

**Comment by msimerson** [comment]
I've wandered in that maze once or twice and I always come out horrified. If I were doing outbound, I'd nuke them tests and just start over. Write some unit tests that verify that each class/file (config, HmailItem, pool, etc) behaves mostly as you expect, and you'll have much better than we've got now.

File: outbound/hmail.js [hunk 1]
```diff
@@ -23,6 +23,10 @@ var client_pool = require('./client_pool');
 var cfg         = require('./config');
 var _qfile      = require('./qfile');
 var queuelib    = require('./queue');
+var mx_lookup   = require('./mx_lookup');
+var outbound    = require('./index');
+
+var get_tls_options = require('./_get_tls_options');
```
Comment (hatsebutz) [inline outbound/hmail.js]: Can not see the file _get_tls_options - any hints?

**Comment by baudehlo** [comment]
Sorry about that - must have forgot to add it. I'll try and do it tonight.

**Comment by baudehlo** [comment]
I've added that file and fixed a test. Still failing. Is node_unit just hiding the real error somewhere? Any way to get it to be more verbose?

**Comment by msimerson** [comment]
> Is node_unit just hiding the real error somewhere? Any way to get it to be more verbose?

You are experiencing the joy that is test/fixtures/vmharness, aka vm.runInNewContext, and the hell it imposes on anyone attempting to debug code that's run in *other* contexts. You're experiencing why I'm such an advocate for getting rid of it.

**Comment by baudehlo** [comment]
How can I atone for this sin I committed? ;)

**Comment by hatsebutz** [comment]
I've converted 2 of the tests using vm_harness for Bounce-Triggering into plain node_unit. How to get them included here, or should I just make a new PR?

**Comment by msimerson** [comment]
@hatsebutz, you can create a PR against this branch (outbound-split).

**Comment by hatsebutz** [comment]
PR #1889 against outbound-split reworks the failing tests from outbound_protocol from using vm_harness to plain node_unit. I've moved them out of the folder one level up and renamed them.

**Comment by msimerson** [comment]
To fix the outbound.js conflicts that prevented merging, I have rebased this against master. I also added a commit fixing the path to fsync_writestream. Because outbound is already broken in master, I'm +1 on merging this and getting it some wider testing.

**Comment by baudehlo** [comment]
There's a few things that are out of date in it. I need to bring it up to where outbound.js currently is in master first. Unfortunately as discussed before, that's a manual process.

**Comment by msimerson** [comment]
I realize it's not quite done, but to make it easier for baudehlo to progress, and b/c outbound in master is already broken, I'm merging.

**Comment by baudehlo** [comment]
OK and it's a good prod for me anyway... Now I just have to figure out how
to get the outbound.js just before this merge ;)

**Comment by msimerson** [comment]
> Now I just have to figure out how to get the outbound.js just before this merge ;)

Easy, github file history browser.

**Comment by msimerson** [comment]
[link](https://github.com/haraka/Haraka/commits/master/outbound.js)

**Pull request merged**